### ISSUE TITLE
why is gulp-rev-all in the blacklist?

### ISSUE THREAD

**Issue by damoncasale** [gulpjs/plugins#219]
I'm resubmitting this issue, based on a comment from @contra that discussions on this should be open.  I sincerely hope that will be the case.

I'm another user who struggled with gulp-rev before discovering gulp-rev-all.  Upon reading the discussion [here](https://github.com/gulpjs/plugins/issues/215) it appears as though @phated is or was responsible for maintaining the blacklist, so I'd like to add my two cents in that firstly, I see a difference between the two plugins as well, and secondly, @phated commented that gulp-rev-all had "poor architecture"  -- and that was apparently one of the other reasons for rejecting it, after @circlingthesun clearly showed how gulp-rev and gulp-rev-all were different, but failed to explain exactly what he meant by that.

Any thoughts by other users of gulp-rev-all?

**Comment by phated** [comment]
Still looks like poor architecture.  You shouldn't need to construct a new object to have access to the stream methods.

> but failed to explain exactly what he meant by that.

It's a waste of time for us to teach people how to implement correct architecture in node modules 😖

**Comment by damoncasale** [comment]
That may be true, and thank you for at least listing some of the issues, but for those of us users who don't feel like delving into the code to see what the architectural issues might be, we rely on the Blacklist as a supposedly UNBIASED list of duplicated plugins.  And it seems like gulp-rev-all is NOT a duplicate, for all of the legitimate reasons that @circlingthesun pointed out.  But rather than acknowledge that, the goalpost was moved to "poor architecture" instead.

Granted, poor architecture may be a legitimate issue, but that should be submitted as a ticket on gulp-rev-all, not included as a reason for listing gulp-rev-all on the Blacklist.

Did you submit any of these architectural issues to their repo, btw?

**Comment by phated** [comment]
No.  Again, architecture audits and guidance is not a service provided by the gulp team.

The reason for blacklisting should be updated.  Sorry that it's not a priority because we have a laundry list of features to get finished for gulp 4.

**Comment by contra** [comment]
@phated Poor performance/efficiency isn't a reason for something to be blacklisted. No plugin codebase is perfect and we certainly don't require that every single thing be written with 100% efficient.

We certainly can't go around critiquing efficiency and code cleanliness when our current release looks like this: https://github.com/robrich/orchestrator/blob/master/index.js

@damoncasale As long as it isn't violating anything [here](https://github.com/gulpjs/plugins#blacklisting) I'll remove it from the blacklist, but after reviewing it myself, I agree with @phated about the construction of the plugin - you shouldn't be instantiating classes in tasks to use the plugin. Fix that and I'll remove it from the list.

cc @smysnk

**Comment by damoncasale** [comment]
It's not my plugin, it's someone else's.  But it seems rather condescending on @phated's part to say, on the one hand, that architecture isn't a reason to be blacklisted, but then implied that there's some _other_ reason it's still blacklisted but won't say what that is.

@contra, thank you for being understanding, though.

**Comment by phated** [comment]
Re-reading the thread you linked, both architecture problems were brought up there and were even discussed by the author yet never addressed.  Pretty absurd for you to claim that I never explained the position. 😖

**Comment by damoncasale** [comment]
I didn't find the issues you were talking about in my (admittedly cursory) review of gulp-rev-all.  I only saw the discussion (linked in my first comment here) where the presence of architectural issues were raised but never specifically spelled out.  But you _did_ spell them out in your initial reply here, which I thanked you for.

Anyway, let's see if the gulp-rev-all maintainer makes the modifications now.

**Comment by phated** [comment]
@damoncasale they were all spelled out in that issue you linked in the first comment and the author even contributed to the thread acknowledging the changes could/should be made (they never were).

**Comment by damoncasale** [comment]
Maybe I'm reading this differently:

@circlingthesun There are two phases. First you determine the dependency graph, the after you know this you can compute file hashes and do replacements. If myapp.js is piped through the plugin, you need the hashes of everything it references before you can compute its hash and pass it downstream. I'm not sure how you would do that without buffering every file.

@phated sounds like further thought needs to go into it...

@circlingthesun I was hoping that you would enlighten us. I'm genuinely curious as to how the architecture could be improved. I would argue that that 1000+ large files don't all need to be searched for references. In which case you probably don't have to buffer those ones that are only hashed and not searched. Maybe pdf's or large video files. Or did you have something else in mind?

Also what "standard node module practices" are being violated?

There was no reply from you after that, that I can see.  So I'm not sure what you're referring to.

**Comment by phated** [comment]
@damoncasale I can see how "standard node module practices" can be misinterpreted or confusing. Better terminology would be what I wrote here, but same sentiment.  You shouldn't have to construct an object to access streams (that usage goes against almost every other plugin).  There was no need to follow up because the author understood what should be done and commented as such, but again, never implemented.

**Comment by damoncasale** [comment]
Well, thank you for the clarification that this is what would qualify gulp-rev-all for being removed from the blacklist.

**Comment by phated** [comment]
@damoncasale np; however, fixing that problem is a very large architectural change for them.  Hopefully they will fix it.

**Comment by damoncasale** [comment]
I, as an average user, wouldn't know whether a small architectural change is required or a large one.  I had trouble following the specific issues listed in the linked discussion.  I only know that it seemed as if @circlingthesun explained how the code worked and rhetorically asked how it would be possible to re-architect that.  He suggested a possible change, although that change doesn't seem to line up with the change you're talking about.  (I'm not sure, I haven't looked at their code.)

If the issue you're talking about has to do with "buffering every file" as @circlingthesun described, then I'm not sure there's any reasonable way for gulp-rev-all to ever get off of the blacklist.  But again, thank you for clarifying that "poor architecture" is yet another reason for listing a plugin on the Blacklist.  That's important for people to know, when they're determining just how much credence to give the Blacklist.  It might also help to actually list WHICH listed reason any Gulp plugin met to have ended up on the Blacklist.

**Comment by smysnk** [comment]
Actually I am not sure I completely understood these "standard node module practices" discussed previously.  First it was about gulp-rev-all duplicating gulp-rev, then discussion moved on to combing through the code base which I am sure other non-blacklisted plugins had the privilege to be exposed to extreme vetting.

So let me see if I understand.. this is to say gulp-rev-all is not allow to maintain state between .revision() and .versionFile() methods?  or this state needs to be stored in a non-neweable gulp-rev-all object.

```
var gulp = require('gulp');
var revAll = require('gulp-rev-all');

gulp.task('default', function () {
  return gulp.src(['assets/**'])
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.revision())
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.versionFile())
    .pipe(gulp.dest('build/assets'));
});
```
instead of..
```
var gulp = require('gulp');
var RevAll = require('gulp-rev-all');

gulp.task('default', function () {

  var revAll = new RevAll();
  return gulp.src(['assets/**'])
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.revision())
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.versionFile())
    .pipe(gulp.dest('build/assets'));

});
```

**Comment by phated** [comment]
It's allowed to maintain state but it should do it without an explicit `new` from the user. The way most other plugins do it. Many store information on the vinyl objects.

**Comment by phated** [comment]
I still think buffering until flush is wrong because you are passing the file along the stream but aren't done with it (as per the guidelines); however, @contra seems to think differently.

**Comment by smysnk** [comment]
State is used to keep options which wouldn't make sense to keep on the vinyl objects.  I suppose these options could possibly be passed into the .revision() method but would make internal architecture more awkward.

Is this allowed?
```
var revision = require('gulp-rev-all')({ /*options*/ });
```
Because we currently support this syntax since May.

**Comment by phated** [comment]
No. The exported object should be a stream returning function or an object with stream returning functions. It should not be a constructor or constructor-like method.

**Comment by smysnk** [comment]
Seems like a super silly reason to get this plugin off the blacklist.  I want to be able to new up multiple revisioners with different configurations which your requirement does not support.  I am not changing the plugin over this requirement.  Sorry.

**Comment by damoncasale** [comment]
What about not using a constructor, but instead using a "settings" function to set data on the vinyl object, to accomplish the same thing?  Then you wouldn't need multiple constructors.  Any setting not set by the settings function would revert to a default setting, rather than any previous setting from before.

How does that sound?

**Comment by phated** [comment]
That doesn't make sense.  Everyone else just accepts options in the stream returning functions.  Not breaking user assumptions is not a silly reason 😖

**Comment by phated** [comment]
gulp-rev-all isn't some special unicorn.  Patterns used by other plugins can be applied in the same way...

**Comment by smysnk** [comment]
Keeping settings in a vinyl object will work, will post back here when I have it implemented.

**Comment by phated** [comment]
Seems like a generated unique ID and a module-local object would be better but anything is better than current.

**Comment by contra** [comment]
@smysnk What's wrong with this? This is how every plugin works, and how every user expects plugins to work:

```js
var gulp = require('gulp');
var revAll = require('gulp-rev-all');

gulp.task('default', function () {
  return gulp.src(['assets/**'])
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.revision({ /* options */ }))
    .pipe(gulp.dest('build/assets'))
    .pipe(revAll.versionFile({ /* options */ }))
    .pipe(gulp.dest('build/assets'));
});
```

It isn't a silly reason at all - we have guidelines about what makes something a plugin. One of those guidelines is to have an API that looks like this, with any state between plugins existing on the file objects. If your module doesn't have this kind of an API then it isn't a plugin, shouldn't be tagged `gulpplugin`, and goes on the blacklist.

Please let me know if this is unclear and needs further explanation, but I think we're beating a dead horse at this point.

**Comment by smysnk** [comment]
I don't think it makes sense to keep a global setup on individual file objects.  Do I just keep a copy of global setup on each file object even though the context doesn't make sense?  When I want to have the options, I must first access the first file?  Extremely silly.

So let me get this right? A gulp plugin isn't a plugin because it keeps state in a central object.

Really you guys come across very combative and I don't feel like I really care to contribute to your ecosystem with attitudes like this. Somebody can fork my plugin and try and work with you guys because I don't care to.

**Comment by damoncasale** [comment]
Diplomats they're not, although I find @contra to be more open to things.

I think a good part of the difficulty is in explaining what they need in a way you can understand.  My impression of @phated's earlier comment is that when .revision is called, (or .settings if you need to set global settings before calling .revision), if the object hasn't been initialized yet, then initialize it and create a module-local object to hold the settings and address it by a unique ID which is retained in a static value, so it won't get confused with later settings.

Yes, especially @phated's attitude leaves a lot to be desired.  But in the end I think they're trying to keep things standardized for the community.  I think that's a worthwhile goal, so if there's anything I can do to encourage you to persevere anyway, despite the bad experience you've had with getting your plugin blacklisted, let me know.  I haven't even looked at your code myself, yet, but I'm willing to take a stab at it if that'll help.

**Comment by contra** [comment]
@smysnk 

I really am trying to help get your plugin off the list - seriously. If anyone is being combative it's you, calling us "extremely silly" - you're not being very understanding at all and you're refusing to make this tiny minor change to the API.

I'll try to boil this down to be really simple and non-combative or whatever it is you think I'm doing:

*Problem:* We're trying to keep the ecosystem **cohesive** and easy to use, and your plugin does not expose an API similar to other plugins.

*Solution:* All you have to do is remove that `new` keyword, because plugins are supposed to be usable without instantiation. You can put the state wherever you feel is easiest (files, global, whatever), we don't care - that isn't the problem here. Just remove the instantiation so your plugin works like every other plugin.

It isn't silly at all, it's a really basic requirement that has been in our guidelines since day 1. Functions that return streams, it's the premise of this entire ecosystem.

**Comment by damoncasale** [comment]
@contra I think he got that way after getting frustrated dealing with @phated in the initial thread I linked before.  Yes, calling requirements 'silly' is non-productive, but one wins more friends with honey than with vinegar.  That's what I'm attempting to do, here.

**Comment by si14** [comment]
I would also add that [the line in the blacklist](https://github.com/gulpjs/plugins/blob/master/src/blackList.json#L70) says `"duplicate of gulp-rev"`. The exchange in this thread makes it clear that the plugin is *not* a duplicate, despite being a fork and serving a similar reason. I believe this is the thing we all can agree on.

**Comment by contra** [comment]
@si14 That reasoning WAS true when it was originally blacklisted, then it was updated a bunch and now it does different things. The current reason is different than why it was originally added, so that should probably be updated. PR?

**Comment by ernestoalejo** [comment]
Sorry if it was already proposed in other issue, but what about an API similar to `gulp-typescript`?:

```js
var gulp = require('gulp');
var revAll = require('gulp-rev-all');

gulp.task('default', function () {
  var r = revAll.create();

  return gulp.src(['assets/**'])
    .pipe(gulp.dest('build/assets'))
    .pipe(r.revision())
    .pipe(gulp.dest('build/assets'))
    .pipe(r.versionFile())
    .pipe(gulp.dest('build/assets'));
});
```

**Comment by si14** [comment]
Following our [discussion on twitter](https://twitter.com/contrahacks/status/776467084612108288): dear @contra, you've stated that

>theres a defined ruleset of what makes something a "plugin" thats existed since day 1 

Can I ask you to point me to the particular rule in [the guidelines](https://github.com/gulpjs/gulp/blob/master/docs/writing-a-plugin/guidelines.md) that is currently violated by the package, so I can create a PR?

It would also be *really* nice to have an answer to @ernestoalejo's question, because it will either mean that `gulp-typescript` should be blacklisted, or that `gulp-rev-all` can be trivially fixed.

**Comment by contra** [comment]
@si14 https://github.com/gulpjs/gulp/blob/master/docs/writing-a-plugin/guidelines.md#what-does-a-good-plugin-look-like

I've added additional info to clarify this, in case the examples aren't clear enough: https://github.com/gulpjs/gulp/commit/0469094bb8c4c54e5bb8b2cf5994dd72fa8db713

Please let me know if the docs could use any more work to explain this! I'll happily merge any documentation PRs that can provide more helpful info.

@ernestoalejo I'm looking at gulp-typescript and the API doesn't look like your example, are you talking about a different module?

I posted earlier in this thread what I think the solution should look like, but I'll post it again:

```js
var gulp = require('gulp');
var revAll = require('gulp-rev-all');

gulp.task('default', function () {
  return gulp.src(['assets/**'])
    .pipe(gulp.dest('some-dir'))
    .pipe(revAll.revision({ /* options */ }))
    .pipe(gulp.dest('some-dir'))
    .pipe(revAll.versionFile())
    .pipe(gulp.dest('some-dir'));
});
```

The `revision` stream tacks whatever options are needed to each file object that passes through it, so that `versionFile()` (or other plugins!) further down the stream can read it. Not a huge change, right? And now it looks and works like the other 3000 plugins!

**Comment by smysnk** [comment]
https://www.npmjs.com/package/gulp-typescript#incremental-compilation
```Instead of calling ts(options), you can create a project first, and then call ts(project).```
Typescript plugin does in fact violate your strict ad-hoc api guidelines.

It is not a big deal to satisfy these api requirements even though I never saw them clearly communicated anywhere until here (besides your 2hour ago commit to guide-lines).  I understand them now and we're clear and everything is good.

`Not a huge change, right? And now it looks and works like the other 3000 plugins!`
This attitude is unnecessary though.

**Comment by contra** [comment]
@smysnk No attitude intended, just excited - I think this kind of stuff is pretty cool once it clicks.

**Comment by ernestoalejo** [comment]
@contra Sorry, I don't want to say your option is bad. I think it's pretty clear and I'm not aware of the problems that have blocked the change in the plugin.

I wanted to explore if there are plugins out there that need today some kind of non-global but accumulated from several files custom state. The proposed solution makes impossible to run parallel tasks with `gulp-rev-all` in both of them, for example.

**Comment by contra** [comment]
@ernestoalejo

> The proposed solution makes impossible to run parallel tasks with gulp-rev-all in both of them, for example.

Could you provide a quick example that shows a situation where putting state on the files would break?

**Comment by ernestoalejo** [comment]
```js
gulp.task('one-of-many', function () {
  return gulp.src(['assets1/**'])
    .pipe(gulp.dest('some-dir1'))
    .pipe(revAll.revision({ /* options */ }))
    .pipe(gulp.dest('some-dir1'))
    .pipe(revAll.versionFile())
    .pipe(gulp.dest('some-dir1'));
});

gulp.task('one-of-others', function () {
  return gulp.src(['assets2/**'])
    .pipe(gulp.dest('some-dir2'))
    .pipe(revAll.revision({ /* options */ }))
    .pipe(gulp.dest('some-dir2'))
    .pipe(revAll.versionFile())
    .pipe(gulp.dest('some-dir2'));
});

gulp.task('default', ['one-of-others', 'one-of-many']);
```

The files that will end in each version file will depend on the speed of execution that changes with each run.

PS They are not limited to tasks, they could be as well `async.forEach(projects, function(projects) { gulp.src(...) });` inside a single task, or a merge of different streams.

**Comment by ernestoalejo** [comment]
`gulp-cache` for example has to expose the `name` option to avoid this problem.

**Comment by ernestoalejo** [comment]
TLDR; if there are some plugins that need that state; can we offer an unified approach with the current rules or updating the rules? Even if the no-state is preferred, if those plugins exist, better to have a recipe for them than leaving each one to a different implementation (which is precisely the thing you're trying to avoid with the rules in the first place).

**Comment by smysnk** [comment]
API has been modified to conform to guidelines. Thanks to everyone who chimed in here!

**Comment by contra** [comment]
@smysnk LGTM - thanks.
